clustering/__init__.py

```diff
@@ -1,67 +1,12 @@
 import pandas as pd
-import numpy as np
-from sklearn import preprocessing
-from sklearn.decomposition import PCA
 from sklearn.cluster import KMeans
 
+def cluster(pd_rules: pd.DataFrame) -> dict:
+    """
+    Cluster data
 
-def find_incident(labels):
-    res = set()
-    for i in range(len(labels)):
-        if labels[i].startswith('i_'):
-            res.add(i)
-    return res
-
-
-# numpy array construction
-def construct_np_allin(allin, m, n, incident_labels):
-    np_allin = []
-    for i in range(m):
-        tmp = []
-        count = 0
-        for j in range(n):
-            if allin[i][j] == True:
-                tmp.append(1)
-            elif allin[i][j] == False:
-                tmp.append(0)
-            elif allin[i][j] != 'None':
-                tmp.append(1)
-            else:
-                tmp.append(0)
-            if j in incident_labels and allin[i][j] == 1:
-                count += 1
-        np_allin.append(tmp+[count])
-    return np_allin
-
-
-def cluster(pd_rules):
-    # global incident_labels, m, n, pd_allin
-    np_rules = pd_rules.as_matrix()
-    labels = list(pd_rules.columns.values)
-    allin = np_rules[:, 3:]
-    allin_labels = labels[3:]
-    incident_labels = find_incident(allin_labels)
-    m, n = allin.shape
-    np_allin = construct_np_allin(allin, m, n, incident_labels)
-    np_allin = np.array(np_allin)
-    pd_allin = pd.DataFrame(data=np_allin[:, :-1], columns=allin_labels)
-    # Kmeans clustering model pipeline
-    min_max_scaler = preprocessing.StandardScaler()
-    np_scaled = min_max_scaler.fit_transform(pd_allin)
-    data = pd.DataFrame(np_scaled)
-    pca = PCA()
-    pca.fit(data)
-    pca = PCA(n_components=3)
-    data = pca.fit_transform(data)
-    min_max_scaler = preprocessing.StandardScaler()
-    np_scaled = min_max_scaler.fit_transform(data)
-    data = pd.DataFrame(np_scaled)
-    # Choose the k by using elbow method
-    n_cluster = range(1, 50)
-    kmeans = [KMeans(n_clusters=i).fit(data) for i in n_cluster]
-    # The ultimate output: system with its associated cluster id -- pd_allin['cluster']
-    pd_allin['cluster'] = kmeans[4].predict(data)
-    # pd_allin['principal_feature1'] = data[0]
-    # pd_allin['principal_feature2'] = data[1]
-    pd_allin['system_id'] = pd_rules['system_id']
-    return dict(zip(pd_allin['system_id'], pd_allin['cluster']))
+    Normalize data and than run KMeans on it (categorize into 5 clusters)
+    """
+    data = ~pd_rules.isin([False, 'False', None, 'None', 0])
+    clusters = KMeans(n_clusters=5).fit_predict(data)
+    return dict(zip(pd_rules['system_id'], clusters))
```

Comment:
@tumido , @durandom This all looks fine to me from a DS perspective. 

The only things that I would note are that removing the "find_incident" function is changing the scope of the data being consumed here from just the "incident" records to all data fields (those that have the prefix "c_" and "r_"). Unless we are specifically concerned with "incident" clustering then I think this is fine, but thought it worth mentioning.

Second, if you do include all fields, they are not all boolean or None. Some are text. Your current implementation on line 10 converts any text value that is not None into True. These instances are fairly rare, and might not have too much of an impact on the overall clustering performance. But again, worth mentioning. They could be addressed with some categorical label encoding as a next step.

Reply:
@tumido OK, my mistake. You are right, I thought ```construct_np_allin``` was constructing a dataset of only 'all incidents'. But I see now that's not the case, its still processing all fields. My bad. 

Yes, I agree. Let's experiment a bit, then we can refine from there. But as I said, current updates seem great. Far more streamlined.

Reply:
WOW. I thought the code was not good. But reducing it to 2 lines looks like magic :)

regarding the experimentation: I think we should start thinking about how we can make sure that changes like these dont impact the results/quality of the model.

@tumido we were experimenting with mlflow.org for that matter. have a look and let's discuss that next week